Transport/Measure-EmailDelayInMTL.ps1

```diff
@@ -0,0 +1,172 @@
+
+.DESCRIPTION
+Gather message tracking log details of all message to / from a given recipient for a given time range.
+Recommend using Start-HistoricalSearch in EXO.
+
+The script will provide an output of all unique message ids with the following information:
+MessageID
+Time Sent
+Total Time in transit
+
+Useful for determining if a "slow" message was a one off or a pattern.
+
+.PARAMETER MTLFile
+MTL File to process.
+
+.PARAMETER ReportPath
+Folder path for the output file.
+
+
+.OUTPUTS
+CSV File with the following information.
+    MessageID               ID of the Message
+    TimeSent                First time we see the message in the MTL
+    TimeReceived            Last delivery time in the MTL
+    MessageDelay            How long before the message was delivered
+
+Default Output File:
+$PSScriptRoot\MTL_report.csv
+
+.EXAMPLE
+.\Measure-EmailDelayInMTL -MTLPath C:\temp\MyMtl.csv
+
+Generates a report from the MyMtl.csv file.
+
+#>
+
+[CmdletBinding()]
+param (
+    [Parameter()]
+    [string]
+    $MTLFile,
+    [Parameter()]
+    [string]
+    $ReportPath = $PSScriptRoot
+)
+
+. $PSScriptRoot\..\Shared\ScriptUpdateFunctions\Test-ScriptVersion.ps1
+
+Function Test-CSVData {
+    param(
+        [array]$CSV,
+        [array]$ColumnsToCheck
+    )
+
+    $ColumnHeaders = ($CSV | Get-Member -MemberType NoteProperty).Name
+    ForEach ( $ColumnToCheck in $ColumnsToCheck) {
+        If ($ColumnHeaders.Contains($ColumnToCheck) ) {
+            # Nothing to do.
+        } Else {
+            Return $false
+        }
+    }
+    Return $true
+}
+
+if (Test-ScriptVersion -AutoUpdate) {
+    # Update was downloaded, so stop here.
+    Write-Host "Script was updated. Please rerun the command."
+    return
+}
+
+# make sure out output variable is null
+$output = $Null
+
+# Test for the provided file and load it.
+if (Test-Path $MTLFile) {
+    Write-Output "Loading MTL file."
+    $mtl = Import-Csv $MTLFile -Encoding Unicode
+} else {
+    Write-Error "Unable to find specified file"
+}
+
+# Validate the MTL
+if (Test-CSVData -CSV $mtl -ColumnsToCheck "event_id", "source", "message_id", "date_time_utc") {}
+else {
+    Write-Error "MTL is missing one or more required fields: `"event_id`",`"source`",`"message_id`",`"date_time_utc`"" -ErrorAction Stop
+}
+
+# get all of the unique message IDs in the file.
+[array]$uniqueMessageIDs = $mtl | Select-Object -ExpandProperty message_id | Sort-Object | Get-Unique
+
+if ($uniqueMessageIDs.count -eq 0) {
+    Write-Error "No Unique MessageIDs found in data."
+}
+
+# Carve the data up into smaller collections to make searching faster.
+# Most of what is in the MTL we don't need for this.
+$SMTPReceive = $mtl | Where-Object { ($_.event_id -eq 'RECEIVE') -and ($_.source -eq 'SMTP') }
+$StoreDeliver = $mtl | Where-Object { ($_.event_id -eq 'DELIVER') -and ($_.source -eq 'STOREDRIVER') }
+$SMTPDeliver = $mtl | Where-Object { ($_.event_id -eq 'SENDEXTERNAL') -and ($_.source -eq 'SMTP') }
```

Comment:
I would assume it is faster to have this all in a single `foreach` loop and adding the data that you want to lists when matched vs looping through all the data 4 different times to add to individual lists.

Reply:
Most email will only have two events that we care about.
The rest is just garbage for the purpose of this script (~4-8 lines depending)
Doing it this way quickly gets rid of the events that we don't want so that when we filter by messageID we are not reading thru 100s or 1000s of events that are not relevant.

Six of one half dozen the other this is just how it makes sense to do this to me.